include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+class KnobSlider final
+    : public Knob
+{
+public:
+    KnobSlider(const std::string& flagName, T defaultValue, T minValue, T maxValue)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(minValue < maxValue, "invalid range to initialize slider");
+        PPX_ASSERT_MSG(minValue <= defaultValue && defaultValue <= maxValue, "defaultValue is out of range");
+
+        mMinValue = minValue;
+        mMaxValue = maxValue;
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::SliderInt(mDisplayName.c_str(), &mValue, mMinValue, mMaxValue, NULL, ImGuiSliderFlags_AlwaysClamp);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <mMinValue~mMaxValue> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <" + std::to_string(mMinValue) + "~" + std::to_string(mMaxValue) + ">";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <int>
+    void UpdateFromFlag(const CliOptions& opts)
+    {
+        SetDefault(opts.GetExtraOptionValueOrDefault(mFlagName, mValue));
+    }
+
+    T GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Updates mValue
+    void SetValue(T newVal)
+    {
+        if (!IsValidValue(newVal)) {
+            PPX_LOG_ERROR(mFlagName << " cannot be set to " << newVal << " because it's out of range " << mMinValue << "~" << mMaxValue);
+            return;
+        }
+
+        mValue = newVal;
+    }
+
+private:
+    // IsValidValue returns true if val is within allowed range
+    bool IsValidValue(T val)
+    {
+        if (val < mMinValue || val > mMaxValue)
+            return false;
+        return true;
+    }
+
+    // SetDefault will set mDefaultValue and update mValue
+    void SetDefault(T newVal)
+    {
+        PPX_ASSERT_MSG(IsValidValue(newVal), "invalid default value");
+        mDefaultValue = newVal;
+        ResetToDefault();
+    }
+
+private:
+    T mValue;
+    T mDefaultValue;
+
+    // mValue will be clamped to the mMinValue to mMaxValue range, inclusive
+    T mMinValue;
+    T mMaxValue;
+};
+
+// KnobDropdown will be displayed as a dropdown in the UI
+// The knob stores the index of a selected choice from a list of allowed options
+// Currently supports option variants of type: string
+template <typename T, std::enable_if_t<std::is_same_v<T, std::string>, bool> = true>
+class KnobDropdown final
+    : public Knob
+{
+public:
+    KnobDropdown(const std::string& flagName, size_t defaultIndex, std::vector<T>::iterator choicesBegin, std::vector<T>::iterator choicesEnd)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(choicesBegin != choicesEnd, "too few allowed choices");
```

Comment:
This check is redundant with the one below:
if distance is 0, then defaultIndex >= size